### PR TITLE
SMILES_PE

### LINKED ISSUE #49: Integrate SMILES-PE
We should integrate the tokenizers:

https://github.com/XinhaoLi74/SmilesPE

### PR DESCRIPTION
**Pull request by jannisborn**
Fixes #49 

Added the SMILES Pair Encoding as described in this paper:

[Li, Xinhao, and Denis Fourches. "SMILES Pair Encoding: A Data-Driven Substructure Tokenization Algorithm for Deep Learning." (2020).](https://chemrxiv.org/articles/SMILES_Pair_Encoding_A_Data-Driven_Substructure_Tokenization_Algorithm_for_Deep_Learning/12339368)

Building upon their package `SmilesPE` ([here](https://github.com/XinhaoLi74/SmilesPE)).

**PR Summary**:
- Included a method called `spe_smiles_tokenizer` in `pytoda.smiles.processing`. This splits SMILES into substructures of varying length, depending on occurrence of the substructures on a corpus of 3.4M SMILES from ChEMBL. E.g. 
```py
spe_smiles_tokenizer( 'c1ccc(/C=C/[C@H](C)O)cc')
>>> ['c1ccc(', '/C=C/', '[C@H](C)', 'O)', 'cc']
```
- Included a method called `kmer_smiles_tokenizer` in `pytoda.smiles.processing`. This splits SMILES into substructures of length `k` where `k` can be set. Additional hyperparameter is `stride`. Note that `k=1` and `stride=1` collapses to the case of our current tokenizer.

```py
kmer_smiles_tokenizer( 'c1ccc(/C=C/[C@H](C)O)cc', k=2, stride=2)
>>> ['[O-][n+]', '1c', 'cc', 'cc', '1S']
kmer_smiles_tokenizer( 'c1ccc(/C=C/[C@H](C)O)cc', k=2, stride=1)
>>> ['c1', '1c', 'cc', 'cc', 'c(', '(/', '/C', 'C=', '=C', 'C/', '/[C@H]', '[C@H](', '(C', 'C)', ')O', 'O)', ')c', 'cc']
kmer_smiles_tokenizer( 'c1ccc(/C=C/[C@H](C)O)cc', k=4, stride=4)
['c1cc', 'c(/C', '=C/[C@H]', '(C)O'], ['[O-][n+]1c', 'cccc']
```

- Both tokenizers can be easily passed to a `SMILESLanguage` object whenever it is instantiated.
E.g.
```py
spe_language = SMILESLanguage(smiles_tokenizer=spe_smiles_tokenizer)
kmer_language = SMILESLanguage(
    smiles_tokenizer=lambda x: kmer_smiles_tokenizer(x, k=2, stride=2)
)
```

- Included `SmilesPE` in `pytoda` dependency
- Added unittests for both `kmer_smiles_tokenizer` and `spe_smiles_tokenizer`

@C-nit Since you're refactoring `pytoda` a lot currently, could you please review?  I also added you @niladell since this could be powerful to condense explicit SMILES representations of polymers which could be a baseline for the paper ;)

### CODE REVIEW

**Comment by jannisborn** [comment]
Btw, @drugilsberg this PR was super easy, thanks to the good infrastructure we developed back in the time when designing the smiles languages! 👍

**Review by niladell — APPROVED** [review]
It looks good to me.

Maybe I'll try to have some models trained with BPE and see how they perform compared to the pubchem model we already have.

File: setup.py [hunk 1]
```diff
@@ -25,7 +25,7 @@
     license='MIT',
     install_requires=[
         'numpy', 'scikit-learn', 'pandas', 'torch>=1.0.0', 'diskcache', 'dill',
-        'selfies', 'upfp'
+        'selfies', 'upfp', 'SmilesPE==0.0.3'
```
Comment (C-nit) [inline setup.py]:
```suggestion
        'selfies', 'upfp', 'SmilesPE'
```
I assume something in version 0.0.4 was not working? If it is just broken and might be fixed in later versions I would stay lenient on the version restriction (in the install_requires). Or did the API change alot and we are sure not to allow any past/future versions?

**Review by C-nit — CHANGES_REQUESTED** [review]
Using the metadata file like this works only in a development environment.
Any distribution will leave the file behind, resulting in:
`FileNotFoundError: [Errno 2] No such file or directory: '.../site-packages/pytoda/smiles/metadata/spe_chembl.txt'`

I can contribute a fix for this.

### TODO:  
The SMILESPE is under an Apache Licence, so we have to give attribution somewhere (for the txt).

File: setup.py (hunk 1 above)
Comment (jannisborn) [reply setup.py]: Interesting. I think the state of the repo has deviated from the `pypi` packge ([repo](https://github.com/XinhaoLi74/SmilesPE/blob/master/SmilesPE/__init__.py)). If you do a plain `pip install` you get `0.0.3`, this is why I fixed it to that one (wasn't even aware a newer one exists). But probably better to be lenient, I agree.

Comment (jannisborn) [reply setup.py]: I made all dependencies lenient `>=0.0.3` @C-nit.

**Comment by jannisborn** [comment]
> Using the metadata file like this works only in a development environment.
> Any distribution will leave the file behind, resulting in:
> `FileNotFoundError: [Errno 2] No such file or directory: '.../site-packages/pytoda/smiles/metadata/spe_chembl.txt'`
> 
> I can contribute a fix for this.
> 
--> if you can fix this, that would be fantastic!

**Comment by drugilsberg** [comment]
I think the apache and the mit are compatible so not sure we need to.
Regarding the metadata we can solve it by using `pkg_resources` and adding the relevant files to the package data.

**Comment by C-nit** [comment]
`importlib.resources` is the way to go from 3.7 on. Tested with `pip install .` running the test file from the install location.

We had to copy the .txt as it's not in the distribution of SMILESPE. This way of redistributing might require more explicit attribution.

**Comment by jannisborn** [comment]
@C-nit I gave acks to the SMILES PE developers. Now it should be fine to include their `.txt` file natively into our package. If they add the file to their pypi package we can consider refactoring, but since they are not responding, I would move forward here and merge.

**Review by C-nit — APPROVED** [review]

**Pull request merged**